### PR TITLE
[mypyc] Constant integer registers optimization

### PR DESCRIPTION
**Pull request by TH3CHARLie**
Recent changes to mypyc introduce low-level, inline integer ops, which makes both IR and generated C code verbose and potentially hurting performance. This PR introduces an on-demand optimization pass that find out all the registers with constant integer values.

Relates mypyc/mypyc#749 mypyc/mypyc#746

### CODE REVIEW

**Comment by TH3CHARLie** [comment]
for this program:
```python
def f(x: int, y: int) -> bool:
    z = x > y
    w = (x + 100) < y
    return z and w
```

originally: 62 lines of C
```c
char CPyDef_f(CPyTagged cpy_r_x, CPyTagged cpy_r_y) {
    char cpy_r_r0;
    char cpy_r_z;
    CPyTagged cpy_r_r1;
    CPyTagged cpy_r_r2;
    char cpy_r_r3;
    int64_t cpy_r_r4;
    int64_t cpy_r_r5;
    int64_t cpy_r_r6;
    char cpy_r_r7;
    int64_t cpy_r_r8;
    int64_t cpy_r_r9;
    int64_t cpy_r_r10;
    char cpy_r_r11;
    char cpy_r_r12;
    char cpy_r_r13;
    char cpy_r_r14;
    char cpy_r_w;
    char cpy_r_r15;
CPyL0: ;
    cpy_r_r0 = CPyTagged_IsGt(cpy_r_x, cpy_r_y);
    cpy_r_z = cpy_r_r0;
    cpy_r_r1 = 200;
    cpy_r_r2 = CPyTagged_Add(cpy_r_x, cpy_r_r1);
    cpy_r_r4 = 1;
    cpy_r_r5 = cpy_r_r2 & cpy_r_r4;
    cpy_r_r6 = 0;
    cpy_r_r7 = cpy_r_r5 == cpy_r_r6;
    cpy_r_r8 = 1;
    cpy_r_r9 = cpy_r_y & cpy_r_r8;
    cpy_r_r10 = 0;
    cpy_r_r11 = cpy_r_r9 == cpy_r_r10;
    cpy_r_r12 = cpy_r_r7 & cpy_r_r11;
    if (cpy_r_r12) {
        goto CPyL1;
    } else
        goto CPyL2;
CPyL1: ;
    cpy_r_r13 = cpy_r_r2 < cpy_r_y;
    CPyTagged_DecRef(cpy_r_r2);
    cpy_r_r3 = cpy_r_r13;
    goto CPyL3;
CPyL2: ;
    cpy_r_r14 = CPyTagged_IsLt_(cpy_r_r2, cpy_r_y);
    CPyTagged_DecRef(cpy_r_r2);
    cpy_r_r3 = cpy_r_r14;
    goto CPyL3;
CPyL3: ;
    cpy_r_w = cpy_r_r3;
    if (cpy_r_z) {
        goto CPyL5;
    } else
        goto CPyL4;
CPyL4: ;
    cpy_r_r15 = cpy_r_z;
    goto CPyL6;
CPyL5: ;
    cpy_r_r15 = cpy_r_w;
    goto CPyL6;
CPyL6: ;
    return cpy_r_r15;
}
```

with this PR: 52 lines
```c
char CPyDef_f(CPyTagged cpy_r_x, CPyTagged cpy_r_y) {
    char cpy_r_r0;
    char cpy_r_z;
    CPyTagged cpy_r_r2;
    char cpy_r_r3;
    int64_t cpy_r_r5;
    char cpy_r_r7;
    int64_t cpy_r_r9;
    char cpy_r_r11;
    char cpy_r_r12;
    char cpy_r_r13;
    char cpy_r_r14;
    char cpy_r_w;
    char cpy_r_r15;
CPyL0: ;
    cpy_r_r0 = CPyTagged_IsGt(cpy_r_x, cpy_r_y);
    cpy_r_z = cpy_r_r0;
    cpy_r_r2 = CPyTagged_Add(cpy_r_x, 200);
    cpy_r_r5 = cpy_r_r2 & 1;
    cpy_r_r7 = cpy_r_r5 == 0;
    cpy_r_r9 = cpy_r_y & 1;
    cpy_r_r11 = cpy_r_r9 == 0;
    cpy_r_r12 = cpy_r_r7 & cpy_r_r11;
    if (cpy_r_r12) {
        goto CPyL1;
    } else
        goto CPyL2;
CPyL1: ;
    cpy_r_r13 = cpy_r_r2 < cpy_r_y;
    CPyTagged_DecRef(cpy_r_r2);
    cpy_r_r3 = cpy_r_r13;
    goto CPyL3;
CPyL2: ;
    cpy_r_r14 = CPyTagged_IsLt_(cpy_r_r2, cpy_r_y);
    CPyTagged_DecRef(cpy_r_r2);
    cpy_r_r3 = cpy_r_r14;
    goto CPyL3;
CPyL3: ;
    cpy_r_w = cpy_r_r3;
    if (cpy_r_z) {
        goto CPyL5;
    } else
        goto CPyL4;
CPyL4: ;
    cpy_r_r15 = cpy_r_z;
    goto CPyL6;
CPyL5: ;
    cpy_r_r15 = cpy_r_w;
    goto CPyL6;
CPyL6: ;
    return cpy_r_r15;
}
```

Both immediate value in python code and used for inline int ops are optimized out.

**Comment by TH3CHARLie** [comment]
@JukkaL so far I haven't added this to pretty IR.

Also, after some thinking, I think we should better still save the literal value for `LoadInt` even if it's tagged, the double value conversion is pretty backend related and if we convert that so early, the IR would be confusing and some other backend that do not represents integers that way may have to do unnecessary work to convert them back

**Comment by TH3CHARLie** [comment]
Also, case 2(checking constant value involved in assign) may need some kind of dataflow analysis.

File: mypyc/transform/integer_optimization.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from typing import List, Dict
+
+from mypyc.ir.ops import BasicBlock, LoadInt, Assign
+from mypyc.ir.rtypes import is_short_int_rprimitive, is_int_rprimitive
+
+
+def find_constant_integer_registers(blocks: List[BasicBlock],
+                                    convert_tagged: bool = False) -> Dict[str, int]:
+    """
```
Comment (JukkaL) [inline mypyc/transform/integer_optimization.py:9]: Style nit: move the first sentence to the same line as `"""` (e.g. `"""Find all ...`).

File: mypyc/transform/integer_optimization.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from typing import List, Dict
```
Comment (JukkaL) [inline mypyc/transform/integer_optimization.py:1]: Can you rename this to `mypyc.analysis.const_int`? `integer_optimization` doesn't describe what this does, but why, and we generally name things based on "what".

**Comment by TH3CHARLie** [comment]
I'll run the tests and change them once #9163 is merged

File: mypyc/test/test_emitfunc.py [hunk 3]
```diff
@@ -325,7 +330,7 @@ def test_simple(self) -> None:
         fn = FuncIR(FuncDecl('myfunc', None, 'mod', FuncSignature([self.arg], int_rprimitive)),
                     [self.block], self.env)
         emitter = Emitter(EmitterContext(NameGenerator([['mod']])))
-        generate_native_function(fn, emitter, 'prog.py', 'prog')
+        generate_native_function(fn, emitter, 'prog.py', 'prog', False)
```
Comment (msullivan) [inline mypyc/test/test_emitfunc.py:333]: Pass `False` by keyword; boolean arguments are hard to understand without it

**Pull request merged**